Mac Catalyst/ARM64 crash: * Assertion: should not be reached at /Users/runner/work/1/s/src/mono/mono/mini/aot-runtime.c:4109

### Description

Assertion in Mono code when using certain code sequences.


### Reproduction Steps

Repro:

1. Create a Mac Catalyst project, make sure TargetFramework=net8.0-maccatalyst
2. Use this C# code:

```cs
using System;

namespace MySimpleApp {
	public class Program {
		static int Main (string [] args)
		{
			ArrayFromHandleFunc (DefaultConvert);
			Console.WriteLine ($"Success!");

			GC.KeepAlive (typeof (NSObject)); // prevent linking away the platform assembly

			return args.Length;
		}

		static object DefaultConvert (HandleStruct handle)
		{
			return null;
		}

		static public void ArrayFromHandleFunc (Func<HandleStruct, object> createObject)
		{
			Console.WriteLine ($"A");
			createObject (default (HandleStruct));
			Console.WriteLine ($"B");
		}
	}

	public struct HandleStruct {
		IntPtr handle;
	}
}
```

3. Build with .NET 8 (8.0.100-preview.6.23315.27) for ARM64:

```
dotnet build /p:RuntimeIdentifier=maccatalyst-arm64
```

Complete project file:

```xml
<?xml version="1.0" encoding="utf-8"?>
<Project Sdk="Microsoft.NET.Sdk">
	<PropertyGroup>
		<TargetFramework>net8.0-maccatalyst</TargetFramework>
		<OutputType>Exe</OutputType>
		<ApplicationId>MonoAssertion</ApplicationId>
	</PropertyGroup>
</Project>
```

### Expected behavior

No crash.

### Actual behavior

```
2023-06-16 12:30:49.511 MySimpleApp[30575:11687045] Microsoft.MacCatalyst: Invalid IDE Port: -1
2023-06-16 12:30:49.516 MySimpleApp[30575:11687021] Microsoft.MacCatalyst: Debugger not loaded (disabled).
2023-06-16 12:30:49.516 MySimpleApp[30575:11687021] Microsoft.MacCatalyst: Profiler not loaded (disabled)
2023-06-16 12:30:49.888 MySimpleApp[30575:11687021] A
2023-06-16 12:30:49.888 MySimpleApp[30575:11687021] error: * Assertion: should not be reached at /Users/runner/work/1/s/src/mono/mono/mini/aot-runtime.c:4109

=================================================================
	Native Crash Reporting
=================================================================
Got a SIGABRT while executing native code. This usually indicates
a fatal error in the mono runtime or one of the native libraries
used by your application.
=================================================================

=================================================================
	Native stacktrace:
=================================================================
	0x1136a0f68 - /Users/rolf/work/maccore/mono-master/xamarin-macios/tests/dotnet/MySimpleApp/MacCatalyst/bin/Debug/net8.0-maccatalyst/maccatalyst-arm64/MySimpleApp.app/Contents/MonoBundle/libmonosgen-2.0.dylib : mono_dump_native_crash_info
	0x1136549b4 - /Users/rolf/work/maccore/mono-master/xamarin-macios/tests/dotnet/MySimpleApp/MacCatalyst/bin/Debug/net8.0-maccatalyst/maccatalyst-arm64/MySimpleApp.app/Contents/MonoBundle/libmonosgen-2.0.dylib : mono_handle_native_crash
	0x1139b758c - /Users/rolf/work/maccore/mono-master/xamarin-macios/tests/dotnet/MySimpleApp/MacCatalyst/bin/Debug/net8.0-maccatalyst/maccatalyst-arm64/MySimpleApp.app/Contents/MonoBundle/libmonosgen-2.0.dylib : sigabrt_signal_handler.cold.1
	0x1136a081c - /Users/rolf/work/maccore/mono-master/xamarin-macios/tests/dotnet/MySimpleApp/MacCatalyst/bin/Debug/net8.0-maccatalyst/maccatalyst-arm64/MySimpleApp.app/Contents/MonoBundle/libmonosgen-2.0.dylib : mono_runtime_setup_stat_profiler
	0x1a8f7aa24 - /usr/lib/system/libsystem_platform.dylib : _sigtramp
	0x1a8f4bc28 - /usr/lib/system/libsystem_pthread.dylib : pthread_kill
	0x1a8e59ae8 - /usr/lib/system/libsystem_c.dylib : abort
	0x112bb018c - /Users/rolf/work/maccore/mono-master/xamarin-macios/tests/dotnet/MySimpleApp/MacCatalyst/bin/Debug/net8.0-maccatalyst/maccatalyst-arm64/MySimpleApp.app/Contents/MonoBundle/libxamarin-dotnet-debug.dylib : _ZL12log_callbackPKcS0_S0_iPv
	0x1137376d4 - /Users/rolf/work/maccore/mono-master/xamarin-macios/tests/dotnet/MySimpleApp/MacCatalyst/bin/Debug/net8.0-maccatalyst/maccatalyst-arm64/MySimpleApp.app/Contents/MonoBundle/libmonosgen-2.0.dylib : monoeg_g_logv_nofree
	0x11373778c - /Users/rolf/work/maccore/mono-master/xamarin-macios/tests/dotnet/MySimpleApp/MacCatalyst/bin/Debug/net8.0-maccatalyst/maccatalyst-arm64/MySimpleApp.app/Contents/MonoBundle/libmonosgen-2.0.dylib : monoeg_assertion_message
	0x1137377ec - /Users/rolf/work/maccore/mono-master/xamarin-macios/tests/dotnet/MySimpleApp/MacCatalyst/bin/Debug/net8.0-maccatalyst/maccatalyst-arm64/MySimpleApp.app/Contents/MonoBundle/libmonosgen-2.0.dylib : monoeg_log_default_handler
	0x113642c08 - /Users/rolf/work/maccore/mono-master/xamarin-macios/tests/dotnet/MySimpleApp/MacCatalyst/bin/Debug/net8.0-maccatalyst/maccatalyst-arm64/MySimpleApp.app/Contents/MonoBundle/libmonosgen-2.0.dylib : mono_aot_get_trampoline_full
	0x113644fc8 - /Users/rolf/work/maccore/mono-master/xamarin-macios/tests/dotnet/MySimpleApp/MacCatalyst/bin/Debug/net8.0-maccatalyst/maccatalyst-arm64/MySimpleApp.app/Contents/MonoBundle/libmonosgen-2.0.dylib : decode_patches
	0x11363f590 - /Users/rolf/work/maccore/mono-master/xamarin-macios/tests/dotnet/MySimpleApp/MacCatalyst/bin/Debug/net8.0-maccatalyst/maccatalyst-arm64/MySimpleApp.app/Contents/MonoBundle/libmonosgen-2.0.dylib : init_method
	0x113640684 - /Users/rolf/work/maccore/mono-master/xamarin-macios/tests/dotnet/MySimpleApp/MacCatalyst/bin/Debug/net8.0-maccatalyst/maccatalyst-arm64/MySimpleApp.app/Contents/MonoBundle/libmonosgen-2.0.dylib : load_method
	0x11363fff4 - /Users/rolf/work/maccore/mono-master/xamarin-macios/tests/dotnet/MySimpleApp/MacCatalyst/bin/Debug/net8.0-maccatalyst/maccatalyst-arm64/MySimpleApp.app/Contents/MonoBundle/libmonosgen-2.0.dylib : mono_aot_get_method
	0x1135bdb78 - /Users/rolf/work/maccore/mono-master/xamarin-macios/tests/dotnet/MySimpleApp/MacCatalyst/bin/Debug/net8.0-maccatalyst/maccatalyst-arm64/MySimpleApp.app/Contents/MonoBundle/libmonosgen-2.0.dylib : jit_compile_method_with_opt
	0x1135bd828 - /Users/rolf/work/maccore/mono-master/xamarin-macios/tests/dotnet/MySimpleApp/MacCatalyst/bin/Debug/net8.0-maccatalyst/maccatalyst-arm64/MySimpleApp.app/Contents/MonoBundle/libmonosgen-2.0.dylib : mono_jit_compile_method
	0x113657814 - /Users/rolf/work/maccore/mono-master/xamarin-macios/tests/dotnet/MySimpleApp/MacCatalyst/bin/Debug/net8.0-maccatalyst/maccatalyst-arm64/MySimpleApp.app/Contents/MonoBundle/libmonosgen-2.0.dylib : mono_delegate_trampoline
	0x109392d24 - /Users/rolf/work/maccore/mono-master/xamarin-macios/tests/dotnet/MySimpleApp/MacCatalyst/bin/Debug/net8.0-maccatalyst/maccatalyst-arm64/MySimpleApp.app/Contents/MacOS/MySimpleApp : generic_trampoline_delegate
	0x10a9c76b4 - /Users/rolf/work/maccore/mono-master/xamarin-macios/tests/dotnet/MySimpleApp/MacCatalyst/bin/Debug/net8.0-maccatalyst/maccatalyst-arm64/MySimpleApp.app/Contents/MacOS/MySimpleApp : MySimpleApp_Program_ArrayFromHandleFunc_System_Func_2_MySimpleApp_HandleStruct_object
	0x10a9c7470 - /Users/rolf/work/maccore/mono-master/xamarin-macios/tests/dotnet/MySimpleApp/MacCatalyst/bin/Debug/net8.0-maccatalyst/maccatalyst-arm64/MySimpleApp.app/Contents/MacOS/MySimpleApp : MySimpleApp_Program_Main_string__
	0x1092b54a8 - /Users/rolf/work/maccore/mono-master/xamarin-macios/tests/dotnet/MySimpleApp/MacCatalyst/bin/Debug/net8.0-maccatalyst/maccatalyst-arm64/MySimpleApp.app/Contents/MacOS/MySimpleApp : wrapper_runtime_invoke_object_runtime_invoke_dynamic_intptr_intptr_intptr_intptr
	0x1135c21d8 - /Users/rolf/work/maccore/mono-master/xamarin-macios/tests/dotnet/MySimpleApp/MacCatalyst/bin/Debug/net8.0-maccatalyst/maccatalyst-arm64/MySimpleApp.app/Contents/MonoBundle/libmonosgen-2.0.dylib : mono_jit_runtime_invoke
	0x1137c2fe8 - /Users/rolf/work/maccore/mono-master/xamarin-macios/tests/dotnet/MySimpleApp/MacCatalyst/bin/Debug/net8.0-maccatalyst/maccatalyst-arm64/MySimpleApp.app/Contents/MonoBundle/libmonosgen-2.0.dylib : mono_runtime_invoke_checked
	0x1137cb3bc - /Users/rolf/work/maccore/mono-master/xamarin-macios/tests/dotnet/MySimpleApp/MacCatalyst/bin/Debug/net8.0-maccatalyst/maccatalyst-arm64/MySimpleApp.app/Contents/MonoBundle/libmonosgen-2.0.dylib : do_exec_main_checked
	0x113615858 - /Users/rolf/work/maccore/mono-master/xamarin-macios/tests/dotnet/MySimpleApp/MacCatalyst/bin/Debug/net8.0-maccatalyst/maccatalyst-arm64/MySimpleApp.app/Contents/MonoBundle/libmonosgen-2.0.dylib : mono_jit_exec
	0x112bc429c - /Users/rolf/work/maccore/mono-master/xamarin-macios/tests/dotnet/MySimpleApp/MacCatalyst/bin/Debug/net8.0-maccatalyst/maccatalyst-arm64/MySimpleApp.app/Contents/MonoBundle/libxamarin-dotnet-debug.dylib : xamarin_main
	0x10aa867dc - /Users/rolf/work/maccore/mono-master/xamarin-macios/tests/dotnet/MySimpleApp/MacCatalyst/bin/Debug/net8.0-maccatalyst/maccatalyst-arm64/MySimpleApp.app/Contents/MacOS/MySimpleApp : main
	0x1a8bf3f28 - /usr/lib/dyld : start

=================================================================
	Basic Fault Address Reporting
=================================================================
Memory around native instruction pointer (0x1a8f14724):0x1a8f14714  ff 0f 5f d6 c0 03 5f d6 10 29 80 d2 01 10 00 d4  .._..._..)......
0x1a8f14724  03 01 00 54 7f 23 03 d5 fd 7b bf a9 fd 03 00 91  ...T.#...{......
0x1a8f14734  65 e0 ff 97 bf 03 00 91 fd 7b c1 a8 ff 0f 5f d6  e........{...._.
0x1a8f14744  c0 03 5f d6 70 0a 80 d2 01 10 00 d4 03 01 00 54  .._.p..........T

=================================================================
	Managed Stacktrace:
=================================================================
	  at <unknown> <0xffffffff>
	  at <unknown> <0xffffffff>
	  at MySimpleApp.Program:ArrayFromHandleFunc <0x000f3>
	  at MySimpleApp.Program:Main <0x0012f>
	  at System.Object:runtime_invoke_dynamic <0x00127>
=================================================================
make: *** [run-bare] Abort trap: 6
```

### Regression?

Yes.

This started showing up in 8.0.100-preview.6.23315.27 for us (in this maestro [bump](https://github.com/xamarin/xamarin-macios/pull/18441)), but was previously hidden because of https://github.com/dotnet/runtime/issues/87277, so this happened sometime between [8.0.100-preview.6.23279.5 and 8.0.100-preview.6.23315.27](https://github.com/dotnet/installer/compare/2de0392d49...bb4e4357e52dc343e19d65e779cddaf8637ca513)

### Known Workarounds

_No response_

### Configuration

```
$ dotnet --info
.NET SDK:
 Version:   8.0.100-preview.6.23315.27
 Commit:    bb4e4357e5

Runtime Environment:
 OS Name:     Mac OS X
 OS Version:  13.4
 OS Platform: Darwin
 RID:         osx.13-arm64
 Base Path:   /Users/rolf/work/maccore/mono-master/xamarin-macios/builds/downloads/dotnet-sdk-8.0.100-preview.6.23315.27/sdk/8.0.100-preview.6.23315.27/

.NET workloads installed:
 [macos]
   Installation Source: SDK 8.0.100-preview.6
   Manifest Version:    13.3.8601-ci.darc-net8-0-acc169ae-baa9-44f2-9253-c1d0338a7ae2/8.0.100-preview.6
   Manifest Path:       /Users/rolf/work/maccore/mono-master/xamarin-macios/builds/downloads/dotnet-sdk-8.0.100-preview.6.23315.27/sdk-manifests/8.0.100-preview.6/microsoft.net.sdk.macos/WorkloadManifest.json
   Install Type:        FileBased
 [ios]
   Installation Source: SDK 8.0.100-preview.6
   Manifest Version:    16.4.8601-ci.darc-net8-0-acc169ae-baa9-44f2-9253-c1d0338a7ae2/8.0.100-preview.6
   Manifest Path:       /Users/rolf/work/maccore/mono-master/xamarin-macios/builds/downloads/dotnet-sdk-8.0.100-preview.6.23315.27/sdk-manifests/8.0.100-preview.6/microsoft.net.sdk.ios/WorkloadManifest.json
   Install Type:        FileBased
 [maccatalyst]
   Installation Source: SDK 8.0.100-preview.6
   Manifest Version:    16.4.8601-ci.darc-net8-0-acc169ae-baa9-44f2-9253-c1d0338a7ae2/8.0.100-preview.6
   Manifest Path:       /Users/rolf/work/maccore/mono-master/xamarin-macios/builds/downloads/dotnet-sdk-8.0.100-preview.6.23315.27/sdk-manifests/8.0.100-preview.6/microsoft.net.sdk.maccatalyst/WorkloadManifest.json
   Install Type:        FileBased
 [tvos]
   Installation Source: SDK 8.0.100-preview.6
   Manifest Version:    16.4.8601-ci.darc-net8-0-acc169ae-baa9-44f2-9253-c1d0338a7ae2/8.0.100-preview.6
   Manifest Path:       /Users/rolf/work/maccore/mono-master/xamarin-macios/builds/downloads/dotnet-sdk-8.0.100-preview.6.23315.27/sdk-manifests/8.0.100-preview.6/microsoft.net.sdk.tvos/WorkloadManifest.json
   Install Type:        FileBased

Host:
  Version:      8.0.0-preview.6.23315.4
  Architecture: arm64
  Commit:       6899decce7

.NET SDKs installed:
  8.0.100-preview.6.23315.27 [/Users/rolf/work/maccore/mono-master/xamarin-macios/builds/downloads/dotnet-sdk-8.0.100-preview.6.23315.27/sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.App 8.0.0-preview.6.23315.8 [/Users/rolf/work/maccore/mono-master/xamarin-macios/builds/downloads/dotnet-sdk-8.0.100-preview.6.23315.27/shared/Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 8.0.0-preview.6.23315.4 [/Users/rolf/work/maccore/mono-master/xamarin-macios/builds/downloads/dotnet-sdk-8.0.100-preview.6.23315.27/shared/Microsoft.NETCore.App]

Other architectures found:
  x64   [/usr/local/share/dotnet/x64]
    registered at [/etc/dotnet/install_location_x64]

Environment variables:
  Not set

global.json file:
  /Users/rolf/work/maccore/mono-master/xamarin-macios/global.json

Learn more:
  https://aka.ms/dotnet/info

Download .NET:
  https://aka.ms/dotnet/download
```

### Other information

_No response_

--- a/src/mono/mono/mini/aot-compiler.c
+++ b/src/mono/mono/mini/aot-compiler.c
@@ -9390,6 +9390,7 @@ add_referenced_patch (MonoAotCompile *acfg, MonoJumpInfo *patch_info, int depth)
 			case MONO_PATCH_INFO_VIRT_METHOD:
 			case MONO_PATCH_INFO_GSHAREDVT_METHOD:
 			case MONO_PATCH_INFO_GSHAREDVT_CALL:
+			case MONO_PATCH_INFO_SIGNATURE:
 				tmp.type = patch_type;
 				tmp.data.target = data;
 				break;

--- a/src/mono/mono/mini/aot-runtime.c
+++ b/src/mono/mono/mini/aot-runtime.c
@@ -4097,7 +4097,8 @@ decode_patch (MonoAotModule *aot_module, MonoMemPool *mp, MonoJumpInfo *ji, guin
 			case MONO_PATCH_INFO_DELEGATE_INFO:
 			case MONO_PATCH_INFO_VIRT_METHOD:
 			case MONO_PATCH_INFO_GSHAREDVT_METHOD:
-			case MONO_PATCH_INFO_GSHAREDVT_CALL: {
+			case MONO_PATCH_INFO_GSHAREDVT_CALL:
+			case MONO_PATCH_INFO_SIGNATURE: {
 				MonoJumpInfo tmp;
 				tmp.type = patch_type;
 				if (!decode_patch (aot_module, mp, &tmp, p, &p))

--- a/src/mono/mono/mini/mini-runtime.c
+++ b/src/mono/mono/mini/mini-runtime.c
@@ -1359,6 +1359,7 @@ mono_patch_info_equal (gconstpointer ka, gconstpointer kb)
 		return ji1->data.jit_icall_id == ji2->data.jit_icall_id;
 	case MONO_PATCH_INFO_VIRT_METHOD:
 		return ji1->data.virt_method->klass == ji2->data.virt_method->klass && ji1->data.virt_method->method == ji2->data.virt_method->method;
+	case MONO_PATCH_INFO_SIGNATURE:
 	case MONO_PATCH_INFO_GSHAREDVT_IN_WRAPPER:
 		return mono_metadata_signature_equal (ji1->data.sig, ji2->data.sig);
 	case MONO_PATCH_INFO_GC_SAFE_POINT_FLAG:

